pkg/controllers/management/auth/globalroles/fleetworkspace_binding_handler.go

```diff
@@ -0,0 +1,195 @@
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+// fleetWorkspaceBindingHandler manages Bindings created for the InheritedFleetWorkspacePermissions field.
+type fleetWorkspaceBindingHandler struct {
+	crbClient rbacv1.ClusterRoleBindingController
+	crbCache  rbacv1.ClusterRoleBindingCache
+	grCache   mgmtcontroller.GlobalRoleCache
+	rbClient  rbacv1.RoleBindingController
+	rbCache   rbacv1.RoleBindingCache
+	fwCache   mgmtcontroller.FleetWorkspaceCache
+}
+
+func newFleetWorkspaceBindingHandler(management *config.ManagementContext) *fleetWorkspaceBindingHandler {
+	return &fleetWorkspaceBindingHandler{
+		crbClient: management.Wrangler.RBAC.ClusterRoleBinding(),
+		crbCache:  management.Wrangler.RBAC.ClusterRoleBinding().Cache(),
+		grCache:   management.Wrangler.Mgmt.GlobalRole().Cache(),
+		rbClient:  management.Wrangler.RBAC.RoleBinding(),
+		rbCache:   management.Wrangler.RBAC.RoleBinding().Cache(),
+		fwCache:   management.Wrangler.Mgmt.FleetWorkspace().Cache(),
+	}
+}
+
+// ReconcileFleetWorkspacePermissionsBindings reconciles backing RoleBindings and ClusterRoleBindings created for granting permission
+// to fleet workspaces.
+func (h *fleetWorkspaceBindingHandler) reconcileFleetWorkspacePermissionsBindings(globalRoleBinding *v3.GlobalRoleBinding) error {
+	globalRole, err := h.grCache.Get(globalRoleBinding.GlobalRoleName)
+	if err != nil {
+		return fmt.Errorf("unable to get globalRole: %w", err)
+	}
+	if globalRole.InheritedFleetWorkspacePermissions.WorkspaceVerbs == nil &&
+		globalRole.InheritedFleetWorkspacePermissions.ResourceRules == nil {
+		return nil
+	}
+
+	if err = h.reconcileResourceRulesBindings(globalRoleBinding, globalRole); err != nil {
+		return fmt.Errorf("error reconciling fleet permissions rules: %w", err)
+	}
+
+	if err = h.reconcileWorkspaceVerbsBindings(globalRoleBinding, globalRole); err != nil {
+		return fmt.Errorf("error reconciling fleet workspace verbs: %w", err)
+	}
+
+	return nil
+}
+
+func (h *fleetWorkspaceBindingHandler) reconcileResourceRulesBindings(grb *v3.GlobalRoleBinding, gr *v3.GlobalRole) error {
+	if gr.InheritedFleetWorkspacePermissions.ResourceRules == nil {
+		return nil
+	}
+	fleetWorkspaces, err := h.fwCache.List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("unable to list fleetWorkspaces when reconciling globalRoleBinding %s: %w", grb.Name, err)
+	}
+
+	var returnError error
+	for _, fleetWorkspace := range fleetWorkspaces {
+		if fleetWorkspace.Name == localFleetWorkspace {
+			continue
+		}
+		rb, err := h.rbCache.Get(fleetWorkspace.Name, wrangler.SafeConcatName(grb.Name))
+		if err != nil && !apierrors.IsNotFound(err) {
+			returnError = multierror.Append(returnError, err)
+			continue
+		}
+		desiredRB := backingRoleBinding(grb, gr, fleetWorkspace.Name)
+		if apierrors.IsNotFound(err) {
+			_, err = h.rbClient.Create(desiredRB)
+			if err != nil && !apierrors.IsNotFound(err) {
```

Comment:
I'm not really sure that a create call can produce a not found error. Can you elaborate on why this is here?

Reply:
removed